### PR TITLE
8313422: test/langtools/tools/javac 144 test classes uses com.sun.tools.classfile library

### PR DESCRIPTION
**Pull request by xqoasis**
There are 144 tests to convert to the new Class-File API.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313422](https://bugs.openjdk.org/browse/JDK-8313422): test/langtools/tools/javac 144 test classes uses com.sun.tools.classfile library (**Sub-task** - P4)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15112/head:pull/15112` \
`$ git checkout pull/15112`

Update a local copy of the PR: \
`$ git checkout pull/15112` \
`$ git pull https://git.openjdk.org/jdk.git pull/15112/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15112`

View PR using the GUI difftool: \
`$ git pr show -t 15112`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15112.diff">https://git.openjdk.org/jdk/pull/15112.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15112#issuecomment-1708639903)

### CODE REVIEW

File: test/langtools/tools/javac/modules/JavaBaseTest.java [hunk 1]
```diff
@@ -213,57 +207,46 @@ void createClass(Path base, List<String> mods, String target) throws Exception {
...
         Path modInfo = base.resolve("test-modules").resolve("module-info.class");
         Files.createDirectories(modInfo.getParent());
-        new ClassWriter().write(cf2, modInfo.toFile());
+        Classfile.of().buildTo(modInfo.toFile().toPath(), cm1.thisClass().asSymbol(), cb -> {
+            for (ClassElement ce: cm1) {
+                if (!(ce instanceof ModuleAttribute))
+                    cb.with(ce);
+            }
+            cb.with(modAttr2);
+        });
```
Comment (asotona) [inline test/langtools/tools/javac/modules/JavaBaseTest.java:249]: The same case as in OpenModulesTest

File: test/langtools/tools/javac/modules/OpenModulesTest.java [hunk 2]
```diff
@@ -234,37 +236,25 @@ public void testNonZeroOpensInOpen(Path base) throws Exception {
...
+                                                          module.provides());
+
+        Classfile.of().buildTo(miClass, cm.thisClass().asSymbol(), classBuilder -> {
+            for (ClassElement ce: cm.elementList()) {
+                if (! (ce instanceof ModuleAttribute)) {
+                    classBuilder.with(ce);
+                }
+            }
+            classBuilder.with(newModule);
+        });
```
[the remark this answers is not in the document]
Comment (xqoasis) [reply test/langtools/tools/javac/modules/OpenModulesTest.java:257]: You are right. Will update in the next commit.

File: test/langtools/tools/javac/records/RecordCompilationTests.java [hunk 3]
```diff
@@ -1289,16 +1273,15 @@ record R(int recordComponent) {}
...
             int numberOfFieldRefs = 0;
-            for (final File fileEntry : dir.listFiles()) {
+            for (final File fileEntry : Objects.requireNonNull(dir.listFiles())) {
                 if (fileEntry.getName().endsWith("R.class")) {
-                    ClassFile classFile = ClassFile.read(fileEntry);
-                    for (CPInfo cpInfo : classFile.constant_pool.entries()) {
-                        if (cpInfo instanceof ConstantPool.CONSTANT_Fieldref_info) {
+                    ClassModel classFile = Classfile.of().parse(fileEntry.toPath());
+                    for (int i = 1; i < classFile.constantPool().entryCount(); ++i) {
+                        if (classFile.constantPool().entryByIndex(i) instanceof FieldRefEntry fieldRefEntry) {
```
Comment (asotona) [inline test/langtools/tools/javac/records/RecordCompilationTests.java:1280]: I've identified iteration over CP entries is problematic in the Classfile API yet, see:
https://mail.openjdk.org/pipermail/classfile-api-dev/2023-August/000399.html

Here the int i must be incremented by the entry::width or the iteration will fail on double-slot entries (long and double).

File: test/langtools/tools/javac/diags/CheckResourceKeys.java [hunk 4]
```diff
@@ -490,10 +496,10 @@ JavaFileManager.Location findJavacLocation(JavaFileManager fm) {
      */
     void scan(JavaFileObject fo, Set<String> results) throws IOException {
         try (InputStream in = fo.openInputStream()) {
-            ClassFile cf = ClassFile.read(in);
-            for (ConstantPool.CPInfo cpinfo: cf.constant_pool.entries()) {
-                if (cpinfo.getTag() == ConstantPool.CONSTANT_Utf8) {
-                    String v = ((ConstantPool.CONSTANT_Utf8_info) cpinfo).value;
+            ClassModel cm = Classfile.of().parse(in.readAllBytes());
+            for (int i = 1; i < cm.constantPool().entryCount(); ++i) {
+                if (cm.constantPool().entryByIndex(i) instanceof Utf8Entry entry) {
```
Comment (asotona) [inline test/langtools/tools/javac/diags/CheckResourceKeys.java:501]: The same iteration issue as in the RecordCompilationTests

File: test/langtools/tools/javac/modules/IncubatingTest.java [hunk 5]
```diff
@@ -255,34 +264,15 @@ private void copyJavaBase(Path targetDir) throws IOException {
...
+        ClassModel cm = Classfile.of().parse(classfile);
+        ModuleResolutionAttribute res = ModuleResolutionAttribute.of(resolution_flags);
+
+        Classfile.of().buildTo(classfile, cm.thisClass().asSymbol(), cb -> {
+            for (ClassElement ce : cm.elementList()) {
+                if (!(ce instanceof ModuleResolutionAttribute))
+                    cb.with(ce);
+            }
+            cb.with(res);
+        });
```
Comment (asotona) [inline test/langtools/tools/javac/modules/IncubatingTest.java:276]: This is the same transformation pattern (dropping and then end handler).

File: test/langtools/tools/javac/annotations/typeAnnotations/referenceinfos/Driver.java [hunk 6]
```diff
@@ -144,32 +137,98 @@ private Map<String, TypeAnnotation.Position> expectedOf(Method m) {
...
+            }
+            case NEW -> {
+                p = TypeAnnotation.TargetInfo.ofNewExpr(null);
+            }
+            case RESOURCE_VARIABLE -> {
+                List<TypeAnnotation.LocalVarTargetInfo> table = new ArrayList<>();
+                for (int idx = 0; idx < d.lvarOffset().length; ++idx)
+                    table.add(TypeAnnotation.LocalVarTargetInfo.of(null, null, d.lvarIndex()[idx]));
+                p = TypeAnnotation.TargetInfo.ofResourceVariable(table);
+            }
```
Comment (asotona) [inline test/langtools/tools/javac/annotations/typeAnnotations/referenceinfos/Driver.java:201]: Filling missing labels with null is not ideal.
All such unsupported cases should directly throw an exception to indicate invalid state of the test.
BTW silent acceptance of null labels by TypeAnnotation.TargetInfo factory methods is a bug.

Comment (asotona) [reply test/langtools/tools/javac/annotations/typeAnnotations/referenceinfos/Driver.java:201]: FYI: I've filled [bug #8315541](https://bugs.openjdk.org/browse/JDK-8315541)

**Review by asotona — APPROVED** [review]

**Pull request closed without merging**